lib/blocks.php

```diff
@@ -296,3 +296,219 @@ function gutenberg_enqueue_editor_block_styles_assets() {
 	}
 	add_action( 'enqueue_block_editor_assets', 'gutenberg_enqueue_editor_block_styles_assets' );
 }
+
+/**
+ * Renders the classNames and styles for blocks
+ *
+ * @param string $block_content Output of the current block.
+ * @param array  $block Block Object.
+ * @return string New block output.
+ */
+function gutenberg_experimental_apply_classnames_and_styles( $block_content, $block ) {
+	// Don't filter template part blocks since we filter the blocks in each template
+	// part individually.
+	if ( 'core/template-part' === $block['blockName'] ) {
+		return $block_content;
+	}
+
+	if ( isset( $block['attrs'] ) ) {
+		// Check what style features the block supports.
+		$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+		if ( $block_type ) {
+			$supports = gutenberg_experimental_global_styles_get_supported_styles( $block_type->supports );
+		} else {
+			$supports = array();
+		}
+		// Return early if nothing is supported.
+		if ( sizeof( $supports ) === 0 ) {
+			return $block_content;
+		}
+
+		$colors     = gutenberg_experimental_build_css_colors( $block['attrs'], $supports );
+		$typography = gutenberg_experimental_build_css_typography( $block['attrs'], $supports );
+
+		$extra_classes = array_merge(
+			$colors['css_classes'],
+			$typography['css_classes'],
+			isset( $block['attrs']['className'] ) ? array( $block['attrs']['className'] ) : array(),
+			isset( $block['attrs']['align'] ) ? array( 'has-text-align-' . $block['attrs']['align'] ) : array()
+		);
+		$extra_styles  = (
+			$colors['inline_styles'] ||
+			$typography['inline_styles']
+		) ? esc_attr( $colors['inline_styles'] ) .
+			esc_attr( $typography['inline_styles'] )
+		: '';
+
+		$dom = new DOMDocument( '1.0', 'utf-8' );
+
+		if ( ! $dom->loadHTML( $block_content, LIBXML_HTML_NOIMPLIED | LIBXML_HTML_NODEFDTD | LIBXML_COMPACT ) ) {
```

Comment:
out of curiosity, what does this stuff do?

Reply:
Its our alternative to using regex 🤣  (props to Andy for this one).  It loads the block content on a dom object so we can get/set the attributes to return the updated html.

Reply:
Yeh it's the built in PHP DOM parser.